services-custom/s3-event-notifications/src/main/java/software/amazon/awssdk/eventnotifications/s3/internal/DefaultS3EventNotificationReader.java

```diff
@@ -0,0 +1,294 @@
+        return new S3(configurationId, bucket, object, s3SchemaVersion);
+    }
+
+    private S3Object readObject(JsonNode jsonNode) {
+        Map<String, JsonNode> objectNode = expectObjectOrNull(jsonNode, "object");
+        if (objectNode == null) {
+            return null;
+        }
+        String key = expectStringOrNull(objectNode.get("key"), "key");
+        Long size = expectLong(objectNode.get("size"), "size");
+        String eTag = expectStringOrNull(objectNode.get("eTag"), "eTag");
+        String versionId = expectStringOrNull(objectNode.get("versionId"), "versionId");
+        String sequencer = expectStringOrNull(objectNode.get("sequencer"), "sequencer");
+        return new S3Object(key, size, eTag, versionId, sequencer);
+    }
+
+    private S3Bucket readBucket(JsonNode jsonNode) {
+        Map<String, JsonNode> bucketNode = expectObjectOrNull(jsonNode, "bucket");
+        if (bucketNode == null) {
+            return null;
+        }
+        String name = expectStringOrNull(bucketNode.get("name"), "name");
+        UserIdentity ownerIdentity = readOwnerIdentity(bucketNode.get("ownerIdentity"));
+        String arn = expectStringOrNull(bucketNode.get("arn"), "arn");
+        return new S3Bucket(name, ownerIdentity, arn);
+    }
+
+    private UserIdentity readOwnerIdentity(JsonNode jsonNode) {
+        Map<String, JsonNode> ownerIdentityNode = expectObjectOrNull(jsonNode, "ownerIdentity");
+        if (ownerIdentityNode == null) {
+            return null;
+        }
+        String principalId = expectStringOrNull(ownerIdentityNode.get("principalId"), "principalId");
+        return new UserIdentity(principalId);
+    }
+
+    private ResponseElements readResponseElements(JsonNode jsonNode) {
+        Map<String, JsonNode> responseElementNode = expectObjectOrNull(jsonNode, "responseElements");
+        if (responseElementNode == null) {
+            return null;
+        }
+        String requestId = expectStringOrNull(responseElementNode.get("x-amz-request-id"), "x-amz-request-id");
+        String id2 = expectStringOrNull(responseElementNode.get("x-amz-id-2"), "x-amz-id-2");
+        return new ResponseElements(id2, requestId);
+    }
+
+    private RequestParameters readRequestParameters(JsonNode jsonNode) {
+        Map<String, JsonNode> requestParametersNode = expectObjectOrNull(jsonNode, "requestParameters");
+        if (requestParametersNode == null) {
+            return null;
+        }
+        JsonNode sourceIpAddress = requestParametersNode.get("sourceIPAddress");
+        String sourceIpAddressString = expectStringOrNull(sourceIpAddress, "sourceIPAddress");
+        return new RequestParameters(sourceIpAddressString);
+    }
+
+
+    private String expectStringOrNull(JsonNode node, String name) {
+        if (node == null) {
+            return null;
+        }
+        Validate.isTrue(node.isString(), "'%s' was not a string", name);
+        return node.asString();
+    }
```

Comment:
Everywhere we're calling this, we're doing a `get` on `node` first with the same string as `name`. Should we do the `get` here, to reduce the chance of copy+paste bugs making these strings different?